File: 2413-smallest-number-in-infinite-set/smallest-number-in-infinite-set.py
# Brute Force Code & Optimal Code
class SmallestInfiniteSet:
    def __init__(self):
        # store numbers that have been popped and later added back use a list here to keep the numbers sorted so the smallest added-back number is always at index 0.
        self.st = []
        # smallest number that has never been popped initially, the infinite set contains: 1, 2, 3, 4, 5, ... so the first number we can pop is 1.
        self.currSmallest = 1
    def popSmallest(self):
        # if there are numbers that were previously popped and then added back the smallest one must be returned first.
        if self.st:
            # since st is sorted, the smallest added-back number is at index 0.
            result = self.st[0]
            # remove that number from the added-back list.
            self.st.pop(0)
        else:
            # if no number has been added back return the smallest number that has never been popped before.
            result = self.currSmallest
            # move currSmallest to the next number.
            self.currSmallest += 1
        # return the smallest available number.
        return result
    def addBack(self, num):
        # if num >= currSmallest num has never been popped before,
        # so it is already present in the infinite set.
        if num >= self.currSmallest or num in self.st:
            return
        # add the previously popped number back into the available numbers.
        self.st.append(num)
        # sort the list so that the smallest added-back number stays at index 0.
        self.st.sort()

# Time Complexity : O(Nlog)
# Space Complexity : O(N)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)

File: 2413-smallest-number-in-infinite-set/smallest-number-in-infinite-set.cpp
// Brute Force Code & Optimal Code
class SmallestInfiniteSet {
public:
    // store numbers that have been popped and later added back set keeps these numbers sorted automatically so the smallest added-back number is always at begin().
    set<int> st;
    // smallest number that has never been popped.
    int currSmallest;
    // constructor
    SmallestInfiniteSet() {
        // initially the smallest number that has never been popped is 1.
        currSmallest = 1;
    }
    int popSmallest() {
        // if there are numbers that were previously popped and then added back, the smallest one must be returned first.
        if (!st.empty()) {
            // since set is sorted begin() points to the smallest added-back number.
            int result = *st.begin();
            // remove that number from the set.
            st.erase(st.begin());
            // return the smallest available number.
            return result;
        }
        // if no number has been added back return the smallest number that has never been popped before.
        int result = currSmallest;
        // move currSmallest to the next number.
        currSmallest++;
        // return the smallest available number.
        return result;
    }
    void addBack(int num) {
        // if num >= currSmallest num has never been popped before,
        // so it is already present in the infinite set.
        if (num >= currSmallest || st.find(num) != st.end()) {
            return;
        }
        // add the previously popped number back into the available numbers.
        st.insert(num);
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)